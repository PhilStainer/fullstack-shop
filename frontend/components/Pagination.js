import {useQuery} from '@apollo/react-hooks'
import PropTypes from 'prop-types'
import Head from 'next/head'
import Link from 'next/link'

import ITEMS_CONNECTION_QUERY from '#root/graphql/itemsConnection.query'
import StyledPagination from '#root/components/styles/StyledPagination'

const perPage = 4

const Items = ({page}) => {
  const {data, loading, error} = useQuery(ITEMS_CONNECTION_QUERY)

  const count = data?.itemsConnection?.totalCount
  const pages = Math.ceil(count / perPage)

  if (loading) return <p>Loading...</p>
  if (error)
    return (
      <p>
        Error:
        {error.message}
      </p>
    )

  return (
    <div style={{textAlign: 'center'}}>
      <StyledPagination data-testid="pagination">
        <Head>
          <title>
            Fullstack shop! - Page {page} of {pages}
          </title>
        </Head>

        <Link href={{pathname: '/', query: {page: page - 1}}}>
          <a className="prev" aria-disabled={page <= 1}>
            ← Prev
          </a>
        </Link>

        <p>
          Page {page} of <span className="totalPages">{pages}</span>!
        </p>

        <p>
          {count} Item{count > 1 && 's'} Total
        </p>

        <Link href={{pathname: '/', query: {page: page + 1}}}>
          <a className="next" aria-disabled={page >= pages}>
            Next →
          </a>
        </Link>
      </StyledPagination>
    </div>
  )
}

Items.propTypes = {
  page: PropTypes.number.isRequired,
}

export default Items
