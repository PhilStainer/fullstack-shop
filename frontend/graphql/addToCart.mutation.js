import gql from 'graphql-tag'

const ADD_TO_CART_MUTATION = gql`
  mutation($id: ID!) {
    addToCart(id: $id) {
      _id
      quantity
      item {
        _id
        title
        description
        imageUrl
        price
      }
    }
  }
`

export default ADD_TO_CART_MUTATION
