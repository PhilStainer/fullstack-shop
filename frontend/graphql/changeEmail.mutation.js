import gql from 'graphql-tag'

const CHANGE_PASSWORD_MUTATION = gql`
  mutation($password: String!, $email: String!, $confirmEmail: String!) {
    changeEmail(
      password: $password
      email: $email
      confirmEmail: $confirmEmail
    ) {
      status
      message
    }
  }
`

export default CHANGE_PASSWORD_MUTATION
