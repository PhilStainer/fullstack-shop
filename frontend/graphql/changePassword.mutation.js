import gql from 'graphql-tag'

const CHANGE_PASSWORD_MUTATION = gql`
  mutation(
    $currentPassword: String!
    $password: String!
    $confirmPassword: String!
  ) {
    changePassword(
      currentPassword: $currentPassword
      password: $password
      confirmPassword: $confirmPassword
    ) {
      status
      message
    }
  }
`

export default CHANGE_PASSWORD_MUTATION
