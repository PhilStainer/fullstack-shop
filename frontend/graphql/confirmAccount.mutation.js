import gql from 'graphql-tag'

const CONFIRM_ACCOUNT_MUTATION = gql`
  mutation($confirmToken: String!) {
    confirmAccount(confirmToken: $confirmToken) {
      status
      message
    }
  }
`

export default CONFIRM_ACCOUNT_MUTATION
