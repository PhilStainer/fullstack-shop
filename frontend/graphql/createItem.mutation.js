import gql from 'graphql-tag'

const CREATE_ITEM_MUTATION = gql`
  mutation(
    $title: String!
    $description: String!
    $imageUrl: String!
    $price: Int!
  ) {
    createItem(
      title: $title
      description: $description
      imageUrl: $imageUrl
      price: $price
    ) {
      _id
    }
  }
`

export default CREATE_ITEM_MUTATION
