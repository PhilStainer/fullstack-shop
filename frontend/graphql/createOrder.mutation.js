import gql from 'graphql-tag'

const CREATE_ORDER_MUTATION = gql`
  mutation($token: String!) {
    createOrder(token: $token) {
      _id
    }
  }
`

export default CREATE_ORDER_MUTATION
