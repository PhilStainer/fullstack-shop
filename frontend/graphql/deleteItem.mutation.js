import gql from 'graphql-tag'

const DELETE_ITEM_MUTATION = gql`
  mutation($id: ID!) {
    deleteItem(id: $id) {
      _id
    }
  }
`

export default DELETE_ITEM_MUTATION
