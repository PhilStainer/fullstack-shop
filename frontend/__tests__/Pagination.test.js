import {render, waitFor} from '@testing-library/react'
import {MockedProvider} from '@apollo/react-testing'
import {GraphQLError} from 'graphql'

import Pagination from '#root/components/Pagination'
import ITEMS_CONNECTION_QUERY from '#root/graphql/itemsConnection.query'

test('renders with initial state of loading', async () => {
  const itemsConnection = {
    request: {
      query: ITEMS_CONNECTION_QUERY,
    },
    result: jest.fn(() => ({
      data: {itemsConnection: {totalCount: 1}},
    })),
  }

  const {getByText} = render(
    <MockedProvider mocks={[itemsConnection]} addTypename={false}>
      <Pagination page={1} />
    </MockedProvider>,
  )

  expect(getByText(/loading/i)).toBeInTheDocument()
})

test('render errors', async () => {
  const itemsConnection = {
    request: {
      query: ITEMS_CONNECTION_QUERY,
    },
    result: jest.fn(() => ({
      errors: [new GraphQLError('Error loading item total...')],
    })),
  }

  const {getByText} = render(
    <MockedProvider mocks={[itemsConnection]} addTypename={false}>
      <Pagination page={1} />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(getByText(/error loading/i)).toBeInTheDocument()
  })
})

test('prev and next buttons disabled on 1 page', async () => {
  const itemsConnection = {
    request: {
      query: ITEMS_CONNECTION_QUERY,
    },
    result: jest.fn(() => ({
      data: {itemsConnection: {totalCount: 1}},
    })),
  }

  const {getByTestId} = render(
    <MockedProvider mocks={[itemsConnection]} addTypename={false}>
      <Pagination page={1} />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(getByTestId('pagination')).toMatchSnapshot()
  })
})

test('prev and next buttons enabled with items', async () => {
  const itemsConnection = {
    request: {
      query: ITEMS_CONNECTION_QUERY,
    },
    result: jest.fn(() => ({
      data: {itemsConnection: {totalCount: 20}},
    })),
  }

  const {getByTestId} = render(
    <MockedProvider mocks={[itemsConnection]} addTypename={false}>
      <Pagination page={2} />
    </MockedProvider>,
  )

  await waitFor(() => {
    expect(getByTestId('pagination')).toMatchSnapshot()
  })
})
