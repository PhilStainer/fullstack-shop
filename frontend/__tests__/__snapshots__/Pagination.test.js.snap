// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prev and next buttons disabled on 1 page 1`] = `
<ul
  class="StyledPagination-wh88lk-0 ccfjeE"
  data-testid="pagination"
>
  <a
    aria-disabled="true"
    class="prev"
    href="/?page=0"
  >
    ← Prev
  </a>
  <p>
    Page 
    1
     of 
    <span
      class="totalPages"
    >
      1
    </span>
    !
  </p>
  <p>
    1
     Item
     Total
  </p>
  <a
    aria-disabled="true"
    class="next"
    href="/?page=2"
  >
    Next →
  </a>
</ul>
`;

exports[`prev and next buttons enabled with items 1`] = `
<ul
  class="StyledPagination-wh88lk-0 ccfjeE"
  data-testid="pagination"
>
  <a
    aria-disabled="false"
    class="prev"
    href="/?page=1"
  >
    ← Prev
  </a>
  <p>
    Page 
    2
     of 
    <span
      class="totalPages"
    >
      5
    </span>
    !
  </p>
  <p>
    20
     Item
    s
     Total
  </p>
  <a
    aria-disabled="false"
    class="next"
    href="/?page=3"
  >
    Next →
  </a>
</ul>
`;
